File: code/smv-debugger/src/main/java/org/fbme/smvDebugger/panel/condition/ConditionParser_Test.java
package org.fbme.smvDebugger.panel.condition;

import junit.framework.TestCase;
import junit.framework.Assert;
import java.util.Map;

public class ConditionParser_Test extends TestCase {
  /*package*/ ConditionParser conditionParser;
  public void test_simpleParse() throws Exception {
    Assert.assertEquals("(functionBlock.VAR == TRUE)", conditionParser.parse("functionBlock.VAR == TRUE").toString());
  }
  public void test_complexParse() throws Exception {
    Assert.assertEquals("((functionBlock.VAR1 == TRUE) && (functionBlock.VAR2 == FALSE))", conditionParser.parse("functionBlock.VAR1 == TRUE && functionBlock.VAR2 == FALSE").toString());
  }
  public void test_bracketsParse() throws Exception {
    Assert.assertEquals("((functionBlock.VAR1 == TRUE) && (functionBlock.VAR2 == FALSE))", conditionParser.parse("((functionBlock.VAR1 == TRUE) && (functionBlock.VAR2 == FALSE))").toString());
  }
  public void test_missingRightBracket() throws Exception {
    try {
      conditionParser.parse("(functionBlock.VAR == TRUE");
      Assert.fail();
    } catch (ConditionParseException e) {
      // expected exception 
    }
  }
  public void test_invalidExpression() throws Exception {
    try {
      conditionParser.parse("functionBlock.VAR ==");
      Assert.fail();
    } catch (ConditionParseException e) {
      // expected exception 
    }
  }
  public void test_simpleEvaluate() throws Exception {
    Assert.assertTrue(conditionParser.parse("functionBlock.VAR == TRUE").evaluate(Map.of("functionBlock.VAR", "TRUE")));
  }
  public void test_complexEvaluate() throws Exception {
    Assert.assertTrue(conditionParser.parse("functionBlock.VAR1 == TRUE && functionBlock.VAR2 == FALSE").evaluate(Map.of("functionBlock.VAR1", "TRUE", "functionBlock.VAR2", "FALSE")));
  }
  public void setUp() {
    conditionParser = new ConditionParser();
  }
}

File: code/smv-debugger/src/main/kotlin/org/fbme/smvDebugger/panel/condition/ConditionParser_Test.kt
package org.fbme.smvDebugger.panel.condition

import junit.framework.TestCase
import kotlin.test.assertFailsWith

class ConditionParser_Test : TestCase() {
    lateinit var conditionParser: ConditionParser

    fun test_simpleParse() {
        assertEquals(
            "(functionBlock.VAR == TRUE)",
            conditionParser.parse("functionBlock.VAR == TRUE").toString()
        )
    }

    fun test_complexParse() {
        assertEquals(
            "((functionBlock.VAR1 == TRUE) && (functionBlock.VAR2 == FALSE))",
            conditionParser.parse("functionBlock.VAR1 == TRUE && functionBlock.VAR2 == FALSE").toString()
        )
    }

    fun test_bracketsParse() {
        assertEquals(
            "((functionBlock.VAR1 == TRUE) && (functionBlock.VAR2 == FALSE))",
            conditionParser.parse("((functionBlock.VAR1 == TRUE) && (functionBlock.VAR2 == FALSE))").toString()
        )
    }

    fun test_missingRightBracket() {
        assertFailsWith(ConditionParseException::class) {
            conditionParser.parse("(functionBlock.VAR == TRUE")
        }
    }

    fun test_invalidExpression() {
        assertFailsWith(ConditionParseException::class) {
            conditionParser.parse("functionBlock.VAR ==")
        }
    }

    fun test_simpleEvaluate() {
        assertTrue(
            conditionParser
                .parse("functionBlock.VAR == TRUE")
                .evaluate(mapOf("functionBlock.VAR" to "TRUE"))
        )
    }

    fun test_complexEvaluate() {
        assertTrue(
            conditionParser
                .parse("functionBlock.VAR1 == TRUE && functionBlock.VAR2 == FALSE")
                .evaluate(mapOf("functionBlock.VAR1" to "TRUE", "functionBlock.VAR2" to "FALSE"))
        )
    }

    public override fun setUp() {
        conditionParser = ConditionParser()
    }
}
